File: packages/webui-desktop/src/validation.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { IpcError } from './errors.js';
import { BoundaryReader, readDelimited, readUint32, readUint64, skipField } from './boundary.js';
import type { IpcLimits } from './limits.js';

/** Descriptor metadata emitted by the application generator, not a proto parser. */
export interface MessageShape { readonly fields: readonly FieldShape[] }
export interface FieldShape {
  readonly number: number;
  readonly name: string;
  readonly kind: string;
  readonly message?: number;
  readonly repeated: boolean;
  readonly packed?: boolean;
  readonly optional: boolean;
  readonly oneof?: string;
  readonly mapKey: boolean;
  readonly map?: boolean;
}
const signed64 = new Set(['int64', 'sint64', 'sfixed64']);
const unsigned64 = new Set(['uint64', 'fixed64']);
const signed32 = new Set(['int32', 'sint32', 'sfixed32', 'enum']);
const unsigned32 = new Set(['uint32', 'fixed32']);
const fixed32 = new Set(['fixed32', 'sfixed32', 'float']);
const fixed64 = new Set(['fixed64', 'sfixed64', 'double']);
const utf8 = new TextEncoder();
const utf8Decoder = new TextDecoder('utf-8', { fatal: true });
function invalid(): never { throw new IpcError('invalid-payload'); }
function shapeAt(messages: readonly MessageShape[], index: number | undefined): MessageShape {
  const shape = index === undefined ? undefined : messages[index];
  if (!shape) return invalid();
  return shape;
}

function scalar(value: unknown, kind: string): void {
  if (signed64.has(kind) || unsigned64.has(kind)) {
    if (typeof value !== 'bigint') invalid();
    const signed = signed64.has(kind);
    if (value < (signed ? -0x8000000000000000n : 0n) ||
        value > (signed ? 0x7fffffffffffffffn : 0xffffffffffffffffn)) invalid();
  } else if (signed32.has(kind) || unsigned32.has(kind)) {
    if (typeof value !== 'number' || !Number.isInteger(value)) invalid();
    const signed = signed32.has(kind);
    if (value < (signed ? -2147483648 : 0) || value > (signed ? 2147483647 : 4294967295)) invalid();
  } else if (kind === 'double' || kind === 'float') {
    if (typeof value !== 'number') invalid();
  } else if (kind === 'bytes') {
    if (!(value instanceof Uint8Array)) invalid();
  } else if (kind === 'bool') {
    if (typeof value !== 'boolean') invalid();
  } else if (kind === 'string') {
    if (typeof value !== 'string') invalid();
  } else invalid();
}

/** Validate JS values before a codec can coerce or truncate them. */
export function validateValue(value: unknown, root: number, messages: readonly MessageShape[], limits: IpcLimits): void {
  const stack: { value: unknown; shape: MessageShape; depth: number }[] = [{ value, shape: shapeAt(messages, root), depth: 1 }];
  let entries = 0;
  let scalarBytes = 0;
  const add = (item: unknown, field: FieldShape, depth: number) => {
    if (field.kind === 'message') {
      if (depth >= limits.maxSchemaDepth) invalid();
      stack.push({ value: item, shape: shapeAt(messages, field.message), depth: depth + 1 });
    } else {
      scalar(item, field.kind);
      if (typeof item === 'string') {
        if (item.length > limits.maxFrameBytes) throw new IpcError('payload-too-large');
        scalarBytes += utf8.encode(item).byteLength;
      }
      else if (item instanceof Uint8Array) scalarBytes += item.byteLength;
      if (scalarBytes > limits.maxFrameBytes) throw new IpcError('payload-too-large');
    }
  };
  while (stack.length) {
    const current = stack.pop()!;
    if (!current.value || typeof current.value !== 'object' || Array.isArray(current.value)) invalid();
    const object = current.value as Record<string, unknown>;
    const oneofs = new Set<string>();
    for (const field of current.shape.fields) {
      let item = object[field.name];
      if (field.oneof) {
        if (object[field.name] !== undefined) invalid();
        const union = object[field.oneof] as { $case?: unknown; value?: unknown } | undefined;
        if (union === undefined) continue;
        if (typeof union !== 'object' || union === null) invalid();
        if (!current.shape.fields.some(f => f.oneof === field.oneof && f.name === union.$case)) invalid();
        if (union.$case !== field.name) continue;
        if (oneofs.has(field.oneof) || object[field.name] !== undefined) invalid();
        oneofs.add(field.oneof);
        item = union.value;
      } else if (item === undefined && field.optional) continue;
      if (field.map) {
        if (!(item instanceof Map)) invalid();
        if (current.depth >= limits.maxSchemaDepth) invalid();
        if (entries + item.size > limits.maxCollectionEntriesPerMessage) invalid();
        entries += item.size;
        const entry = shapeAt(messages, field.message);
        const keyField = entry.fields.find(f => f.mapKey || f.number === 1);
        const valueField = entry.fields.find(f => f.number === 2);
        if (!keyField || !valueField) invalid();
        for (const [key, value] of item) {
          scalar(key, keyField.kind);
          if (typeof key === 'string') {
            if (key.length > limits.maxFrameBytes) throw new IpcError('payload-too-large');
            scalarBytes += utf8.encode(key).byteLength;
            if (scalarBytes > limits.maxFrameBytes) throw new IpcError('payload-too-large');
          }
          add(value, valueField, current.depth + 1);
        }
      } else if (field.repeated) {
        if (!Array.isArray(item)) invalid();
        if (entries + item.length > limits.maxCollectionEntriesPerMessage) invalid();
        entries += item.length;
        for (const element of item) add(element, field, current.depth);
      } else add(item, field, current.depth);
    }
  }
}

function wireType(kind: string): number {
  return fixed64.has(kind) ? 1 : fixed32.has(kind) ? 5
    : kind === 'string' || kind === 'bytes' || kind === 'message' ? 2 : 0;
}

function readScalar(reader: BoundaryReader, field: FieldShape): void {
  switch (field.kind) {
    case 'string': utf8Decoder.decode(readDelimited(reader)); return;
    case 'bytes': readDelimited(reader); return;
    case 'bool': {
      const value = readUint32(reader);
      if (value !== 0 && value !== 1) invalid();
      return;
    }
    case 'int64': case 'uint64': case 'sint64': readUint64(reader); return;
    case 'fixed64': reader.fixed64(); return;
    case 'sfixed64': reader.sfixed64(); return;
    case 'int32': case 'enum': {
      const raw = readUint64(reader);
      if (raw > 0x7fffffffn && raw < 0xffffffff80000000n) invalid();
      return;
    }
    case 'uint32': case 'sint32': readUint32(reader); return;
    case 'fixed32': reader.fixed32(); return;
    case 'sfixed32': reader.sfixed32(); return;
    case 'float': reader.float(); return;
    case 'double': reader.double(); return;
    default: invalid();
  }
}

/** Iterative bounded wire guard before generated payload decoding. */
export function validateMessage(bytes: Uint8Array, root: number, messages: readonly MessageShape[], limits: IpcLimits): void {
  if (bytes.byteLength > limits.maxFrameBytes) throw new IpcError('payload-too-large');
  const stack = [{ bytes, shape: shapeAt(messages, root), depth: 1 }];
  let entries = 0;
  try {
    while (stack.length) {
      const current = stack.pop()!;
      const reader = new BoundaryReader(current.bytes);
      const oneofs = new Set<string>();
      while (reader.pos < reader.len) {
        const [number, type] = reader.tag();
        if (type === 3 || type === 4) invalid();
        const field = current.shape.fields.find(f => f.number === number);
        if (!field) { skipField(reader, type); continue; }
        if (field.oneof) {
          if (oneofs.has(field.oneof)) invalid();
          oneofs.add(field.oneof);
        }
        if (field.kind === 'message') {
          if (field.repeated && ++entries > limits.maxCollectionEntriesPerMessage) invalid();
          if (type !== 2 || current.depth >= limits.maxSchemaDepth) invalid();
          stack.push({ bytes: readDelimited(reader), shape: shapeAt(messages, field.message), depth: current.depth + 1 });
        } else if (type === 2 && field.repeated && field.packed === true && wireType(field.kind) !== 2) {
          const packed = new BoundaryReader(readDelimited(reader));
          while (packed.pos < packed.len) {
            if (++entries > limits.maxCollectionEntriesPerMessage) invalid();
            readScalar(packed, field);
          }
        } else {
          if (field.repeated && ++entries > limits.maxCollectionEntriesPerMessage) invalid();
          if (type !== wireType(field.kind)) invalid();
          readScalar(reader, field);
        }
      }
    }
  } catch (error) {
    if (error instanceof IpcError) throw error;
    invalid();
  }
}

File: crates/webui-desktop/src/ipc/validation.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

use super::{
    error::fail,
    wire::{ipc_frame::Body, IpcFrame, Kind},
    IpcError, IpcErrorCode, IpcLimits, IPC_VERSION,
};

/// Generated scalar or message field metadata.
#[derive(Clone, Copy)]
pub enum FieldKind {
    /// Boolean varint.
    Bool,
    /// Signed int32.
    Int32,
    /// Signed int64.
    Int64,
    /// Unsigned int32.
    Uint32,
    /// Unsigned int64.
    Uint64,
    /// Zigzag int32.
    Sint32,
    /// Zigzag int64.
    Sint64,
    /// Unsigned fixed32.
    Fixed32,
    /// Unsigned fixed64.
    Fixed64,
    /// Signed fixed32.
    Sfixed32,
    /// Signed fixed64.
    Sfixed64,
    /// IEEE 32-bit float, including nonfinite values.
    Float,
    /// IEEE 64-bit float, including nonfinite values.
    Double,
    /// Enum numeric value, including unknown values.
    Enum,
    /// Opaque bytes.
    Bytes,
    /// UTF-8 text.
    String,
    /// Index into the generated message table.
    Message(usize),
}
/// One generated message's fields.
pub struct MessageShape {
    /// Schema fields, including map-entry messages.
    pub fields: &'static [FieldShape],
}
/// Generated validation metadata.
pub struct FieldShape {
    /// Protobuf field tag.
    pub number: u32,
    /// Expected wire encoding.
    pub kind: FieldKind,
    /// Whether collection entries must be counted, including packed values.
    pub repeated: bool,
    /// Whether a repeated scalar field accepts packed wire encoding.
    pub packed: bool,
    /// Local oneof group index.
    pub oneof: Option<u32>,
    /// True for the key field in a map-entry message.
    pub map_key: bool,
}

/// Validate wire boundaries before prost decoding, without recursion or allocating
/// declared payload lengths. Unknown fields are bounded and skipped.
pub fn validate_message(
    bytes: &[u8],
    root: usize,
    messages: &'static [MessageShape],
    limits: &IpcLimits,
) -> Result<(), IpcError> {
    if bytes.len() > limits.max_frame_bytes {
        return Err(fail(IpcErrorCode::PayloadTooLarge));
    }
    let shape = messages
        .get(root)
        .ok_or_else(|| fail(IpcErrorCode::SchemaMismatch))?;
    let mut stack = vec![Cursor::new(bytes, shape)];
    let mut entries = 0usize;
    while let Some(cursor) = stack.last_mut() {
        if cursor.input.is_empty() {
            stack.pop();
            continue;
        }
        let (number, wire) = key(&mut cursor.input)?;
        let data = field(&mut cursor.input, wire)?;
        let Some(rule) = cursor.shape.fields.iter().find(|f| f.number == number) else {
            continue;
        };
        if let Some(group) = rule.oneof {
            if cursor
                .oneofs
                .insert(group, number)
                .is_some_and(|old| old != number)
            {
                return Err(fail(IpcErrorCode::InvalidPayload));
            }
        }
        let scalar_wire = expected(rule.kind);
        if rule.repeated && rule.packed && wire == 2 && scalar_wire != 2 {
            let mut packed = data;
            while !packed.is_empty() {
                let value = field(&mut packed, scalar_wire)?;
                validate_scalar(value, rule.kind)?;
                entries += 1;
                if entries > limits.max_collection_entries_per_message {
                    return Err(fail(IpcErrorCode::PayloadTooLarge));
                }
            }
            continue;
        }
        if scalar_wire != wire {
            return Err(fail(IpcErrorCode::InvalidPayload));
        }
        if rule.repeated {
            entries = entries
                .checked_add(1)
                .ok_or_else(|| fail(IpcErrorCode::PayloadTooLarge))?;
            if entries > limits.max_collection_entries_per_message {
                return Err(fail(IpcErrorCode::PayloadTooLarge));
            }
        }
        match rule.kind {
            FieldKind::Message(child) => {
                if stack.len() >= limits.max_schema_depth {
                    return Err(fail(IpcErrorCode::PayloadTooLarge));
                }
                let shape = messages
                    .get(child)
                    .ok_or_else(|| fail(IpcErrorCode::SchemaMismatch))?;
                stack.push(Cursor::new(data, shape));
            }
            FieldKind::String => {
                // Generated browser codecs use Map, not object-backed maps.
                // Every legal protobuf string key remains representable.
                std::str::from_utf8(data).map_err(|_| fail(IpcErrorCode::InvalidPayload))?;
            }
            _ => validate_scalar(data, rule.kind)?,
        }
    }
    Ok(())
}

struct Cursor<'a> {
    input: &'a [u8],
    shape: &'static MessageShape,
    oneofs: std::collections::HashMap<u32, u32>,
}
impl<'a> Cursor<'a> {
    fn new(input: &'a [u8], shape: &'static MessageShape) -> Self {
        Self {
            input,
            shape,
            oneofs: std::collections::HashMap::new(),
        }
    }
}
fn expected(kind: FieldKind) -> u8 {
    match kind {
        FieldKind::Fixed64 | FieldKind::Sfixed64 | FieldKind::Double => 1,
        FieldKind::Fixed32 | FieldKind::Sfixed32 | FieldKind::Float => 5,
        FieldKind::Bytes | FieldKind::String | FieldKind::Message(_) => 2,
        _ => 0,
    }
}
fn validate_scalar(mut bytes: &[u8], kind: FieldKind) -> Result<(), IpcError> {
    if expected(kind) != 0 {
        return Ok(());
    }
    let value = varint(&mut bytes)?;
    let valid = match kind {
        FieldKind::Bool => value <= 1,
        FieldKind::Uint32 | FieldKind::Sint32 => value <= u64::from(u32::MAX),
        FieldKind::Int32 | FieldKind::Enum => {
            value <= i32::MAX as u64 || value >= 0xffff_ffff_8000_0000
        }
        _ => true,
    };
    if !valid {
        return Err(fail(IpcErrorCode::InvalidPayload));
    }
    Ok(())
}

pub(super) fn varint(input: &mut &[u8]) -> Result<u64, IpcError> {
    let mut value = 0u64;
    for shift in (0..70).step_by(7) {
        let Some((&byte, rest)) = input.split_first() else {
            return Err(fail(IpcErrorCode::InvalidFrame));
        };
        *input = rest;
        if shift == 63 && byte > 1 {
            return Err(fail(IpcErrorCode::InvalidFrame));
        }
        value |= u64::from(byte & 127) << shift;
        if byte < 128 {
            return Ok(value);
        }
    }
    Err(fail(IpcErrorCode::InvalidFrame))
}
fn key(input: &mut &[u8]) -> Result<(u32, u8), IpcError> {
    let key = varint(input)?;
    let number = u32::try_from(key >> 3).map_err(|_| fail(IpcErrorCode::InvalidFrame))?;
    if number == 0 || number > 0x1fff_ffff {
        return Err(fail(IpcErrorCode::InvalidFrame));
    }
    let wire = u8::try_from(key & 7).map_err(|_| fail(IpcErrorCode::InvalidFrame))?;
    Ok((number, wire))
}
fn field<'a>(input: &mut &'a [u8], wire: u8) -> Result<&'a [u8], IpcError> {
    let length = match wire {
        0 => {
            let before = *input;
            varint(input)?;
            return Ok(&before[..before.len() - input.len()]);
        }
        1 => 8,
        2 => usize::try_from(varint(input)?).map_err(|_| fail(IpcErrorCode::InvalidFrame))?,
        5 => 4,
        _ => return Err(fail(IpcErrorCode::InvalidFrame)),
    };
    if length > input.len() {
        return Err(fail(IpcErrorCode::InvalidFrame));
    }
    let (data, rest) = input.split_at(length);
    *input = rest;
    Ok(data)
}

/// Decode a strict v2 envelope after checking complete size and wire boundaries.
/// Application bytes remain opaque here; generated validators run on workers.
/// Decode a strict v3 envelope. The fixed layout has no ambiguity to police
/// (no duplicate/unknown fields, no wire-type confusion), so a single decode
/// pass both parses and structurally validates the bytes; only the envelope's
/// own semantic invariants are checked afterward. Application bytes remain
/// opaque here - generated validators run on workers.
pub fn decode_frame(bytes: &[u8], limits: &IpcLimits) -> Result<IpcFrame, IpcError> {
    if bytes.len() > limits.max_frame_bytes {
        return Err(fail(IpcErrorCode::PayloadTooLarge));
    }
    let frame = IpcFrame::decode(bytes).map_err(|_| fail(IpcErrorCode::InvalidFrame))?;
    validate_frame(&frame, limits)?;
    Ok(frame)
}

#[cfg(test)]
mod tests {
    use super::{validate_message, FieldKind, FieldShape, MessageShape};
    use crate::ipc::IpcLimits;

    const REPEATED_INT32: [FieldShape; 1] = [FieldShape {
        number: 1,
        kind: FieldKind::Int32,
        repeated: true,
        packed: false,
        oneof: None,
        map_key: false,
    }];
    const PACKED_INT32: [FieldShape; 1] = [FieldShape {
        packed: true,
        ..REPEATED_INT32[0]
    }];
    const NON_PACKED_MESSAGES: [MessageShape; 1] = [MessageShape {
        fields: &REPEATED_INT32,
    }];
    const PACKED_MESSAGES: [MessageShape; 1] = [MessageShape {
        fields: &PACKED_INT32,
    }];

    #[test]
    fn repeated_scalar_respects_packed_metadata() {
        let packed_bytes = [0x0a, 0x02, 0x01, 0x02];
        assert!(
            validate_message(&packed_bytes, 0, &PACKED_MESSAGES, &IpcLimits::default()).is_ok()
        );
        assert!(validate_message(
            &packed_bytes,
            0,
            &NON_PACKED_MESSAGES,
            &IpcLimits::default()
        )
        .is_err());
    }
}

fn validate_frame(frame: &IpcFrame, limits: &IpcLimits) -> Result<(), IpcError> {
    if frame.version != IPC_VERSION {
        return Err(fail(IpcErrorCode::UnsupportedVersion));
    }
    if frame.generation == 0 || frame.id == 0 {
        return Err(fail(IpcErrorCode::InvalidFrame));
    }
    let kind = Kind::try_from(frame.kind).map_err(|_| fail(IpcErrorCode::InvalidFrame))?;
    let valid = match kind {
        Kind::Request => {
            frame.method_id != 0
                && frame.timeout_ms != 0
                && u64::from(frame.timeout_ms) <= limits.max_timeout_ms as u64
                && matches!(frame.body, Some(Body::Payload(_)))
        }
        Kind::Notify => {
            frame.method_id != 0
                && frame.timeout_ms == 0
                && matches!(frame.body, Some(Body::Payload(_)))
        }
        Kind::Result => {
            frame.method_id == 0
                && frame.timeout_ms == 0
                && matches!(frame.body, Some(Body::Payload(_)))
        }
        Kind::Error => {
            frame.method_id == 0
                && frame.timeout_ms == 0
                && matches!(frame.body, Some(Body::Error(_)))
        }
        Kind::Accept | Kind::Cancel => {
            frame.method_id == 0 && frame.timeout_ms == 0 && frame.body.is_none()
        }
        Kind::Unspecified => false,
    };
    if !valid {
        return Err(fail(IpcErrorCode::InvalidFrame));
    }
    if let Some(Body::Error(error)) = &frame.body {
        let total = error.code.len()
            + error.message.len()
            + error.help.len()
            + error.application_code.len();
        if total > limits.max_error_text_bytes_total {
            return Err(fail(IpcErrorCode::PayloadTooLarge));
        }
    }
    Ok(())
}
